src/Microsoft.Identity.Client/UI/CustomWebUiHandler.cs

```diff
@@ -0,0 +1,107 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Extensibility;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.UI
+{
+    internal class CustomWebUiHandler : IWebUI
+    {
+        private readonly ICustomWebUi _customWebUi;
+
+        public CustomWebUiHandler(ICustomWebUi customWebUi)
+        {
+            _customWebUi = customWebUi;
+        }
+
+        /// <inheritdoc />
+        public async Task<AuthorizationResult> AcquireAuthorizationAsync(
+            Uri authorizationUri,
+            Uri redirectUri,
+            RequestContext requestContext)
+        {
+            requestContext.Logger.Info(LogMessages.CustomWebUiAcquiringAuthorizationCode);
+
+            try
+            {
+                var uri = await _customWebUi.AcquireAuthorizationCodeAsync(authorizationUri, redirectUri)
+                                            .ConfigureAwait(false);
+                if (uri == null)
+                {
+                    throw new MsalCustomWebUiFailedException(CoreErrorMessages.CustomWebUiReturnedNullUri);
+                }
+
+                if (uri.Authority.Equals(redirectUri.Authority, StringComparison.OrdinalIgnoreCase) &&
+                    uri.AbsolutePath.Equals(redirectUri.AbsolutePath))
+                {
+                    IDictionary<string, string> inputQp = CoreHelpers.ParseKeyValueList(
+                        authorizationUri.Query.Substring(1),
+                        '&',
+                        true,
+                        null);
+
+                    requestContext.Logger.Info(LogMessages.CustomWebUiRedirectUriMatched);
+                    return new AuthorizationResult(AuthorizationStatus.Success, uri.OriginalString)
+                    {
+                        State = inputQp[OAuth2Parameter.State]
+                    };
+                }
+
+                throw new MsalCustomWebUiFailedException(CoreErrorMessages.CustomWebUiRedirectUriWasNotMatchedToProperUri);
+            }
+            catch (Exception ex)
+            {
+                var authStatus = AuthorizationStatus.UnknownError;
+
+                if (ex is OperationCanceledException)
+                {
+                    requestContext.Logger.Info(LogMessages.CustomWebUiOperationCancelled);
+                    authStatus = AuthorizationStatus.UserCancel;
+                }
+                else
+                {
+                    requestContext.Logger.WarningPiiWithPrefix(ex, CoreErrorMessages.CustomWebUiAuthorizationCodeFailed);
```

Comment:
For places where we have pii logged, it seem like we are not logging anything about if pii == false, I think we should still log something. Not sure though we can do that properly from the Logger class, it may have to be selected here. thoughts?

Reply:
We log both PII and a scrubbed version:

        public void WarningPiiWithPrefix(Exception exWithPii, string prefix)
        {
            Log(LogLevel.Warning, prefix + exWithPii.ToString(), prefix + MsalExceptionFactory.GetPiiScrubbedExceptionDetails(exWithPii));
        }